LB cleanup

KaiSzuttor opened the pull request:
Fixes #2357. Fixes #2356. Fixes #2491. Fixes #2517

TODOs:

 - [x] coupling seed interface
 - [x] philox for CPU LB fluid
 - [x] fluid seed checkpointing

mkuron commented on CMakeLists.txt:
```diff
@@ -387,12 +387,9 @@ set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/${PYTHON_INSTDIR}/espressomd")
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
-if(CMAKE_VERSION VERSION_LESS 3.1)
-  include(RequireCXX11)
```
You can now delete cmake/RequireCXX11.cmake

mkuron commented on src/core/grid_based_algorithms/lb.cpp:
```diff
@@ -2208,70 +2193,38 @@ void lb_calc_n_from_rho_j_pi(const Lattice::index_t index, const double rho,
+std::array<double, 19> lb_calc_m_from_n(const std::array<double, 19> n) {
+  std::array<double, 19> m;
+  for (int i = 0; i < 19; i++) {
+    m[i] = boost::inner_product(lbmodel.e_ki[i], n, 0.0);
```
Do we lose any efficiency by doing this, or does the `constexpr` suffice?

KaiSzuttor replied:
WIP

mkuron commented on src/core/grid_based_algorithms/lb.cpp:
```diff
@@ -2643,74 +2600,14 @@ inline void lb_viscous_coupling(Particle *p, double force[3]) {
-      if (this_node == 0) {
-        fprintf(stderr, "ERROR: Swimming is not compatible with Open MPI and "
-                        "CPU LB on more than 1 node.\n");
-        fprintf(stderr, "       Please use LB_GPU instead.\n");
```
Is this limitation removed now?

fweik replied:
Yes, as far as I can tell. The existing test passes now with multiple nodes, and conceptually it should work.

mkuron commented on src/core/grid_based_algorithms/lbgpu_cuda.cu:
```diff
@@ -4261,8 +4246,7 @@ void lb_calc_shanchen_GPU() {
 void lb_save_checkpoint_GPU(float *host_checkpoint_vd,
                             unsigned int *host_checkpoint_boundary,
-                            lbForceFloat *host_checkpoint_force,
-                            uint64_t *host_checkpoint_philox_counter) {
```
Where did the counter checkpointing go?

fweik replied:
WIP

fweik commented on CMakeLists.txt:
```diff
@@ -387,12 +387,9 @@ set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/${PYTHON_INSTDIR}/espressomd")
 
 # C++ standard
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 14)
```
I'm all in favour of bumping the standard version to 14, but I think we should do so also
for the CUDA. So you would have to check which CUDA version is needed for 14,
and change that accordingly. Also the documentation has to be adapted in several
places and some code can be removed form utils, because it is now provided by the
language. If this is not strictly needed here, I think it should be a separate PR.

mkuron replied:
CUDA 8 supports C++14. That means dropping support for Ubuntu 16.04's CUDA packages (but one can manually install newer ones).

KaiSzuttor replied:
i think the only reason for this was ```constexpr std::pair``` in utils/u32_to_u64.hpp

mkuron replied:
Ok, that's not enough to justify the switch. You could just do `reinterpret_cast`s between `uint32_t[2]` and `uint64_t` instead of using `std::pair`.

fweik replied:
Just use the `CXX14_CONSTEXPR` macro, it's not important that this function is `constexpr`, it just can be in 14. @mkuron the cleaner fix is to use `std::array` instead of `std::pair`. `array` has aggregate initialization, which is `constexpr` in 11.

RudolfWeeber commented:
Is the newer CUDA available on the bee cluster? 
Otherwise, we probably need to wait for that.

KaiSzuttor commented:
We plan to update bee anyway before Christmas, this PR will anyway be not ready for merge before that.

fweik commented:
CXX14_CONSTEXPR...

On Fri, Nov 30, 2018 at 12:18 PM Kai Szuttor <notifications@github.com>
wrote:

> *@KaiSzuttor* commented on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/espressomd/espresso/pull/2392#discussion_r237827345>:
>
> > @@ -387,12 +387,9 @@ set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/${PYTHON_INSTDIR}/espressomd")
>  #######################################################################
>
>  # C++ standard
> -set(CMAKE_CXX_STANDARD 11)
> +set(CMAKE_CXX_STANDARD 14)
>
> i think the only reason for this was constexpr std::pair in
> utils/u32_to_u64.hpp
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/espressomd/espresso/pull/2392#discussion_r237827345>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2i3wgdlwV-nUB8u83E4BrjQU-V45DFks5u0RP4gaJpZM4Y7SjF>
> .
>


-- 
Florian Weik

florianweik@gmail.com
++49 157 85939252

KaiSzuttor committed:
utils: update of inner_product.

KaiSzuttor committed:
utils: update of inner_product.

KaiSzuttor committed:
LB: fix compiler warning.

KaiSzuttor committed:
LB: fix compiler warning - the right way.

KaiSzuttor committed:
Fix if not LB.

KaiSzuttor committed:
Inline header defined function definition.

KaiSzuttor committed:
LB: LB rho to LB units.

KaiSzuttor committed:
LB: exchange implementation of vector matrix product for mode transformation.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
utils: unit test for matrix_vector_product.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
Bug in gcc 👎  https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52036

KaiSzuttor committed:
LB: workaround gcc bug.

KaiSzuttor committed:
LB: fix units for LB boundaries.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: units.

KaiSzuttor committed:
testsuite: speedup and clean stokes sphere LB test.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
unit_tests: fix matrix vector product for gcc bug.

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: ugly hack to circumvent gcc bug.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into lb_cleanup

KaiSzuttor committed:
EK: Fix unit of LB density.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
unit_tests: braces.

KaiSzuttor committed:
clang-tidy: nolint inheader definition since clang tidy does not see the static keyword coming from the macro.

KaiSzuttor committed:
LB: fix bounceback for shanchen.

KaiSzuttor committed:
Formatting.

fweik committed:
core: lb: Philox for fluid thermalization.

fweik committed:
core: lb: Cleanup

KaiSzuttor committed:
Merge pull request #19 from fweik/lb_cleanup

Phliox for LB CPU fluid thermalization

KaiSzuttor committed:
utils: make_integer_sequence for c++11.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: Initialization braces.

KaiSzuttor committed:
utils: matrix_vector product c++11 compatibility.

KaiSzuttor committed:
LB: include rng state in checkpoint.

KaiSzuttor committed:
LB: counter for fluid rng and require seed.

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
testsuite: missing seeds.

fweik committed:
maintainer: Removed unused parts of Random123

fweik committed:
testsuite: Added missing lb seeds

fweik committed:
lb: Moved interface functions into correct place.

fweik commented:
@KaiSzuttor the remaining build failures are because `std::get` is not `constexpr` in C++11.

KaiSzuttor committed:
CMake: require CUDA 9 and c++14.

KaiSzuttor committed:
CI: remove incompatible builds.

KaiSzuttor committed:
CI: fix dependencies on removed jobs.

KaiSzuttor committed:
CI: fix dependencies on removed jobs.

KaiSzuttor committed:
CMake: cuda version check.

mkuron commented on CMakeLists.txt:
```diff
@@ -143,25 +143,34 @@ if (WITH_CUDA)
+        message(STATUS "Found CUDA version: ${CUDA_VERSION}")
         message(STATUS "Found CUDA installation: ${CUDA_DIR}")
+        string(COMPARE GREATER_EQUAL ${CUDA_VERSION} "9" CUDA_VERSION_IS_VALID)
+        if (NOT ${CUDA_VERSION_IS_VALID})
```
Replace these two lines with `if(CUDA_VERSION VERSION_LESS 9.0 )`.

mkuron commented on CMakeLists.txt:
```diff
@@ -143,25 +143,34 @@ if (WITH_CUDA)
         execute_process(COMMAND ${CUDA_NVCC_EXECUTABLE} ${CMAKE_CXX_FLAGS} --verbose
-                ERROR_VARIABLE CUDA_DIR)
-        string(REGEX REPLACE "^.*Found CUDA installation: ([^,]+).*\$" "\\1" CUDA_DIR "${CUDA_DIR}")
+                ERROR_VARIABLE CUDA_DIR_STRING)
```
fix this line's indentation by adding 9 spaces

KaiSzuttor committed:
CMake: cuda version check.

KaiSzuttor committed:
CMake format.

KaiSzuttor committed:
LB: LB thermostat takes instance of LB fluid. Require seed only if kT>0.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
testsuite: new thermostat syntax.

KaiSzuttor committed:
LB: renaming of thermostats LB fluid parameter.

KaiSzuttor committed:
LB: documentation.

KaiSzuttor committed:
Utils: removed unnecessary integer sequence implementation.

KaiSzuttor committed:
LB: use index sequence from std library.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: some cleanup.

KaiSzuttor committed:
LB: store force density in LB units.

KaiSzuttor committed:
CI: disable cuda for intel 17 due to lack of support for c++14.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: store viscosity in LB units.

KaiSzuttor committed:
LB: store bulk viscosity in LB units.

KaiSzuttor committed:
CI: fix path to script.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
There are several CI containers that have default compiler versions that don't support c++14 fully.

fweik commented:
Yes, gcc 4.8 is out, full support needs gcc 5, 4.9 has most of it though. (According to https://www.gnu.org/software/gcc/projects/cxx-status.html#cxx14)

KaiSzuttor committed:
EK: hack the hacky ek initialization.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
Intel compiler hangs on compiling LB.

https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/62792

fweik commented:
-> https://en.cppreference.com/w/cpp/compiler_support

KaiSzuttor commented:
@fweik @mkuron how do we want to resolve the c++14 issue in centos and opensuse?
And do you have any idea what is wrong with intel17?

KaiSzuttor commented:
> -> https://en.cppreference.com/w/cpp/compiler_support

This needs explanation. @fweik

fweik commented:
What is there to resolve? They have too old compilers, so they can't be supported anymore if we want 14.

KaiSzuttor commented:
there is still the possibility to install newer compilers

mkuron commented:
OpenSuse 42.3: drop, that is going out of support in a few months anyway.
Intel 15: please replace with the oldest Intel compiler that has sufficient C++14 support, e.g. Intel 16.
CentOS 7: please install the oldest possible compiler from the EPEL repository.
Intel 17: unclear, should work. You probably have some unusual template structure that causes the compiler to hang. Also, please upgrade the base image to CUDA 9.0 and then re-enable CUDA.

KaiSzuttor commented:
Concerning intel, i recently read a few posts about various bugs when compiling recursive templates. So it might very well be a bug.

RudolfWeeber commented:
On Fri, Jan 11, 2019 at 07:24:03PM +0000, Michael Kuron wrote:
> OpenSuse 42.3: drop, that is going out of support in a few months anyway.
> Intel 15: please replace with the oldest Intel compiler that has sufficient C++14 support, e.g. Intel 16.
> CentOS 7: please install the oldest possible compiler from the EPEL repository.
> Intel 17: unclear, should work. You probably have some unusual template structure that causes the compiler to hang.
The question of environments we do and don't support is on the
agenda of the next ES meeting.

KaiSzuttor committed:
Merge branch 'python' of github.com:espressomd/espresso into lb_cleanup

KaiSzuttor committed:
CI: remove job for opensuse 42.3.

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into lb_cleanup

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
Partially fix merge regression.

KaiSzuttor committed:
CI: remove intel 17 again.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Fixing merge regression.

KaiSzuttor committed:
cmake: require >icc18.

KaiSzuttor committed:
CI: added intel18.

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
Utils: template type for vector in matrix vector product.

KaiSzuttor committed:
utils: get.

KaiSzuttor committed:
utils: use utils get in matrix vector product.

KaiSzuttor committed:
unit_tests: (re)add test for matrix_vector_product.

KaiSzuttor committed:
LB: introduce reference type for LB fluid.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
Fix merge regression.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: separate the LB core interface.

KaiSzuttor committed:
Merge branch 'python' of github.com:espressomd/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: fix ifdefs.

KaiSzuttor committed:
LB: separate particle coupling for CPU LB.

KaiSzuttor committed:
LB: remove some LB includes.

KaiSzuttor committed:
Merge branch 'python' of github.com:espressomd/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: refactor on_lb_params_change.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: remove occurences of old on_lb_params function.

KaiSzuttor committed:
LB: fix on params change.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: guard interface function that are not yet really interface.

KaiSzuttor committed:
LB: guards.

KaiSzuttor committed:
LB: guards.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: modernization of LB interface and SHANCHEN removed (tests fail).

KaiSzuttor committed:
LB: removed some more SHANCHEN stuff.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: fix gpu LB.

KaiSzuttor committed:
LB: remove remaining SHANCHEN stuff.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
testsuite: fix LB density test.

KaiSzuttor committed:
LB: fix compiler warnings.

KaiSzuttor committed:
LB: else instead of elif.

KaiSzuttor committed:
LB: fix build with empty config.

KaiSzuttor commented:
@fweik, I think you should review this PR for now. We should merge soon, since it's prone to diverge.

fweik commented:
We should involve somebody else for review, probably @RudolfWeeber, also please resolve the merge conflict.

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
doc: fix sphinx build.

KaiSzuttor committed:
CI: fix job dependency.

RudolfWeeber commented:
On Mon, Feb 04, 2019 at 05:56:55AM -0800, Florian Weik wrote:
> We should involve somebody else for review, probably @RudolfWeeber 
Will do, starting Friday.

KaiSzuttor committed:
LB: use local temperature (magically breaks the particle coupling).

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: fix setting of kT.

KaiSzuttor committed:
LB: calculate noise prefactor locally.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: cosmetics.

KaiSzuttor committed:
LB: make reinit_parameters a interface function.

KaiSzuttor committed:
LB: remove fluct from struct.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: some more interface cleanup.

KaiSzuttor committed:
LB: only use LB interface in initialize.

KaiSzuttor committed:
LB: reduce LB guts exposure.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: guard import.

KaiSzuttor committed:
utils: move index ravel to own header.

KaiSzuttor committed:
LB: move bounceback kernel to LB.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: make reinit_fluid an actual interface function.

KaiSzuttor committed:
LB: make initialization a interface function.

KaiSzuttor committed:
LB: make naming consistent.

KaiSzuttor committed:
LB: move fluid statistics to lb.[h,c]pp.

KaiSzuttor committed:
LB: remove unused and untested method.

KaiSzuttor committed:
LB: modernize collect boundary forces.

KaiSzuttor committed:
LB: modernize calc fluid momentum.

KaiSzuttor committed:
LB: move agrid setting logic to lbgpu.cpp.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: remove gut dependency from observables.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Merge branch 'python' of github.com:espressomd/espresso into lb_cleanup

KaiSzuttor committed:
LB: make the coupling an actual interface.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
CI: fix artifacts for additional checks.

KaiSzuttor committed:
utils: unit_test for index.

KaiSzuttor committed:
utils: refactor index.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
inertialess tracers: remove SHANCHEN artifacts.

KaiSzuttor committed:
Merge branch 'python' of github.com:espressomd/espresso into lb_cleanup

RudolfWeeber commented:
```diff
diff --git a/src/core/grid_based_algorithms/lb.cpp b/src/core/grid_based_algorithms/lb.cpp
index 5ec3a0f59..fc7d596e3 100644
--- a/src/core/grid_based_algorithms/lb.cpp
+++ b/src/core/grid_based_algorithms/lb.cpp
@@ -28,9 +28,9 @@
[...]
+  /* setup the initial particle velocity distribution */
+  lb_reinit_fluid();
```

Does that realy have th to do with particles? Please clarify comment

```diff
diff --git a/src/core/grid_based_algorithms/lb_particle_coupling.cpp b/src/core/grid_based_algorithms/lb_particle_coupling.cpp
new file mode 100644
index 000000000..e102cf48d
+  lb_lbfluid_add_force_density(pos, delta_j);
```

Are we really sure, the per-node forces are force densities?
The units don't really suggest that.

RudolfWeeber commented:
Also, my understanding from offline discussion is that
* system.actors.add(lbf) does not add the particle coupling
* The particle coupling is activated separately independent of thermlization
* The fric-parameter belongs to the particle coupling

KaiSzuttor committed:
LB: move friction parameter to particle coupling.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: adapt testsuite to new coupling behavior.

fweik commented on src/core/grid_based_algorithms/lb_particle_coupling.hpp:
```diff
@@ -10,6 +12,36 @@ void lb_lbcoupling_calc_particle_lattice_ia(bool couple_virtual);
+   * noise on the particle coupling and the fluid
+   * thermalization.
+   */
+  enum class RNGSalt { PARTICLES = 2 };
```
This is explicitly _not_ an property of the coupling. Please make this one enum again and put it into a common header. This is so that the compiler generates the numbers and make sure they don't collide.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB interface: guard usage of LB function.

KaiSzuttor committed:
LB: move RNGSalt to random.hpp.

KaiSzuttor committed:
LB: correct false comment.

KaiSzuttor commented:
@RudolfWeeber our force_density is probably more a momentum density since it has to have the dimensions as the modes 1,2,3

KaiSzuttor commented:
The interface changes for the particle coupling are done @fweik @RudolfWeeber

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
LB particle coupling: only noise if kT > 0.0.

KaiSzuttor committed:
LB: propagators.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: remove forgotten counter increments.

KaiSzuttor committed:
LB: high level profiler annotations plus make transfer_momentum a boolean.

KaiSzuttor committed:
LB: fix build with empty.

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
doc: adapt samples to new LB syntax.

KaiSzuttor committed:
LB: remove 3pt coupling because not tested and seems broken.

KaiSzuttor committed:
LB: remove interface function for interpolation order.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
I removed the 3pt coupling because it gives the wrong interpolated velocities.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
@RudolfWeeber @fweik

KaiSzuttor committed:
LB: remove seed from GPU struct.

KaiSzuttor committed:
utils: refactor unravel_index.

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
Fix merge regression.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
cmake: increase test timeout and improve test time for lb test.

RudolfWeeber commented:
In lb_interface.hpp:

please add doxygen to those functions which are not obvious

void lb_lbfluid_propagate();

void lb_lbfluid_update();
What's the difference, and do we need to expose both?



#ifdef LB
extern bool transfer_momentum;
#endif

#ifdef LB_GPU
extern bool transfer_momentum_gpu;
#endif
Should we rename those? They control particle coupling, don't they?


void lb_lbfluid_set_lattice_switch(int local_lattice_switch);
This seems to set more than the switch. Rename to switch_lb_method or the like?


int lb_lbfluid_get_interpolated_velocity_global(Vector3d &p, double *v)
Make v a vector2d?
p could be const
Could the interpolation routines from the private namespae be reused here?

KaiSzuttor committed:
LB: move and rename transfer_momentum to particle coupling.

KaiSzuttor committed:
LB: documentation of the interface.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
CI: timeout for clang tests.

KaiSzuttor commented:
@RudolfWeeber 

I have added docstrings to almost all interface functions.

I now only expose the ```lb_lbfluid_propagate``` method.

I renamed ```transfer_momentum``` to ```couple_to_md``` and added it to the ```LB_Particle_Coupling``` struct.

The ```lb_lbfluid_set_lattice_switch``` method sets a global variable that is used to decide which LB is running. Of course this information is not supposed to be exposed and I will rename the method.

The velocity interpolation function that is not marked as global is actually not yet an interface function, I should probably put some more work into that. The global function can be refactored to use the interpolation stuff of the other. I will do that when I refactor the non-global version.

KaiSzuttor committed:
LB: guard communication.

KaiSzuttor committed:
lattice: refactor map_lattice_to_node.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into lb_cleanup

KaiSzuttor committed:
LB: fix communication of particle coupling.

fweik commented on src/core/lattice.hpp:
```diff
@@ -100,13 +100,12 @@ class Lattice {
    * \return         index of the node for the lattice site
    */
-  int map_lattice_to_node(int *ind, int *grid) const;
+  int map_lattice_to_node(Vector3i &ind) const;
```
Is `ind` an output parameter? Otherwise it should be `const`.

KaiSzuttor replied:
yes

fweik commented on src/python/espressomd/thermostat.pyx:
```diff
@@ -330,39 +343,55 @@ cdef class Thermostat(object):
+            seed=None,
+            act_on_virtual=True,
+            LB_fluid=None,
+                friction=0.0):
```
If think the naming should be consistent with the Langevin thermostat, where the friction parameter is called gamma. Also the docstring for `friction` is missing.

fweik commented on src/python/espressomd/thermostat.pyx:
```diff
@@ -330,39 +343,55 @@ cdef class Thermostat(object):
-            temperature = float(kT)
+                    "seed has to be given as keyword arg")
+
+            if not seed:
```
Can you please check that this logic is actually needed. This is somewhat dangerous because setting the temperature to zero resets the rng state. If you do thermalized -> unthermalized -> thermalized, you get the same noise in the second thermalized round, which probably should not happen.

KaiSzuttor replied:
Otherwise we would have to check if a seed is needed by checking whether ```Utils::Counter::initial_value() != Utils::Counter::value()```, right?

fweik replied:
If think it is enough to just leave the seed alone if the temperature is zero. Otherwise you would eg get an error if you set the thermostat twice without integrating.

KaiSzuttor committed:
CI: fix doxygen check.

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
LB: removed remaining three-point coupling code.

KaiSzuttor committed:
LB: move interpolation functions next to each other.

KaiSzuttor committed:
Lattice: some refactoring.

KaiSzuttor committed:
LB: moved interpolation to separate file.

KaiSzuttor committed:
LB: removed old interpolation function.

KaiSzuttor committed:
LB: added missing files.

KaiSzuttor committed:
LB: interpolation cleanup.

KaiSzuttor committed:
LB interpolation: rename function and clean up of global interpolation method.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Merge branch 'python' of github.com:espressomd/espresso into lb_cleanup

KaiSzuttor committed:
lattice: eliminate remaining occurences of ROUND_ERROR_PREC.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into lb_cleanup

KaiSzuttor committed:
LB interpolation: use folded positions.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
Formatting.

RudolfWeeber commented:
lb.cpp:

```c++

template <typename T>
std::array<T, 19> normalize_modes(const std::array<T, 19> &modes) {
  auto normalized_modes = modes;
  for (int i = 0; i < modes.size(); i++) {
    normalized_modes[i] /= lbmodel.w_k[i];
  }
  return normalized_modes;
}

template <typename T, std::size_t N>
std::array<T, N> lb_calc_n_from_m(const std::array<T, N> &modes) {
  auto ret = Utils::matrix_vector_product<T, N, ::D3Q19::e_ki_transposed>(
      normalize_modes(modes));
  std::transform(ret.begin(), ret.end(), ::D3Q19::w.begin(), ret.begin(),
                 std::multiplies<T>());
  return ret;
}
```

It might be possible to avoid 19 multiplications by storing a combinaiton of e_k_i_transposed or w and 1/w_k (from normalize_modes)


lb_collide_stream()
It might improve performance to do the mode-level operations (deterministic collide, thermalize) in-place. 
No clue wether the compiler does that on its own.

Why is the boundary bounce back happening after the halo comm?


utils/matrix_vector_product.hpp
Please add a generous amount of comments throughout the file.
It should probably go to utils/math

fweik commented:
> Why is the boundary bounce back happening after the halo comm?

Because the halo comm does the streaming, otherwise the populations are not up to date
if a boundary is a halo node.

KaiSzuttor committed:
utils: moved matrix_vector_product to math.

KaiSzuttor committed:
utils: documentation matrix-vector product.

KaiSzuttor commented:
> lb_collide_stream()
It might improve performance to do the mode-level operations (deterministic collide, thermalize) in-place.
No clue wether the compiler does that on its own.

I checked for that and it didn't make a difference.

KaiSzuttor commented:
> utils/matrix_vector_product.hpp
Please add a generous amount of comments throughout the file.
It should probably go to utils/math

The implementation details just contain a recursive template which is kind of standard.
I will document the interface function.

KaiSzuttor commented:
> It might be possible to avoid 19 multiplications by storing a combinaiton of e_k_i_transposed or w and 1/w_k (from normalize_modes)

No, because we have to use an ```int``` matrix for the template magic in the matrix vector product.

KaiSzuttor committed:
LB: friction->gamma.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: do not set seed if no temperature.

fweik committed:
Merge branch 'python' into lb_cleanup

KaiSzuttor committed:
LB: setter for node density and velocity.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into lb_cleanup

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
LB: removed leftover debug prints.

KaiSzuttor committed:
LB interface: fixing build.

KaiSzuttor committed:
LB: fix units.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
observables: fixing LB velocity units.

KaiSzuttor committed:
LB: minor.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
@itischler could you please have a look which changes break the EK? You are probably faster than me ;)

itischler commented:
No Problem. I'll do it

KaiSzuttor commented:
@itischler git bisect says it's commit 4de1a20de0b87d6e18fc10d66e01e300e3e222e2 to blame

itischler committed:
Fixed the scaling of the lb pressure tensor in the LB and EK case.
Also enabled the multiplication of a vector6d with a scalar in the
python interface.

KaiSzuttor committed:
LB: rename function.

KaiSzuttor committed:
LB: interface for pressure.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into itischler_lb_cleanup

KaiSzuttor committed:
Merged espressomd/python.

KaiSzuttor committed:
Merge branch 'lb_cleanup' of github.com:KaiSzuttor/espresso into lb_cleanup

KaiSzuttor committed:
LB: maybe tricked cython.

KaiSzuttor commented:
@fweik

fweik commented:
This is a serious improvement. Overall this looks good to me, this is really good work. There are still some points that should be addressed now or later: 

- The cython classes should get a proper inheritance hierarchy with a common base class. If a new LB is added, say using walberla, it should be obvious how to do that.
- Please double check that the changed build requirements are properly reflected in the documentation and everywhere else (e.g. tutorial 0).
- Move halo and lattice to the lb folder.
- The remaining globals should be eliminated from CPU LB (e.g. so that it can be build as a separate library), e.g. it should take the local box and so on as parameters. This is probably for a later day.

We should consult @RudolfWeeber again and then bring this to a conclusion soon.

KaiSzuttor committed:
LB interface: class hierarchy.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
doc: update compiler requirement.

KaiSzuttor committed:
LB interface: feature guard.

KaiSzuttor commented:
@RudolfWeeber please press the button if you are happy

fweik reviewed (changes requested):
Can you please move the `lattice` and `halo` files into the correct directory?

KaiSzuttor committed:
halo/lattice: moved to grid_based_algorithms.

KaiSzuttor committed:
lattice: correct include path.

fweik commented on src/core/forces.cpp:
```diff
@@ -37,7 +37,7 @@
 #include "grid_based_algorithms/lb_particle_coupling.hpp"
 #include "immersed_boundaries.hpp"
-#include "lattice.hpp"
+#include "grid_based_algorithms/lattice.hpp"
```
Why is this needed here?

fweik commented on src/core/global.cpp:
```diff
@@ -29,7 +29,7 @@
 #include "grid.hpp"
 #include "initialize.hpp"
-#include "lattice.hpp"
+#include "grid_based_algorithms/lattice.hpp"
```
Why is this needed here?

KaiSzuttor replied:
because ```lattice_switch``` is a global variable

fweik commented on src/core/initialize.cpp:
```diff
@@ -48,7 +48,7 @@
 #include "grid_based_algorithms/lb_interface.hpp"
 #include "grid_based_algorithms/lbboundaries.hpp"
-#include "lattice.hpp"
+#include "grid_based_algorithms/lattice.hpp"
```
Why is this needed here?

fweik commented on src/core/integrate.cpp:
```diff
@@ -42,7 +42,7 @@
 #include "grid_based_algorithms/lb_particle_coupling.hpp"
 #include "initialize.hpp"
-#include "lattice.hpp"
+#include "grid_based_algorithms/lattice.hpp"
```
Why is this needed here?

fweik commented on src/core/rotation.cpp:
```diff
@@ -43,7 +43,7 @@
 #include "initialize.hpp"
 #include "integrate.hpp"
-#include "lattice.hpp"
+#include "grid_based_algorithms/lattice.hpp"
```
Why is this needed here?

fweik commented on src/python/espressomd/globals.pxd:
```diff
@@ -56,7 +56,7 @@ cdef extern from "integrate.hpp":
     extern bool set_py_interrupt
 
-cdef extern from "lattice.hpp":
+cdef extern from "grid_based_algorithms/lattice.hpp":
```
Why is this needed here?

KaiSzuttor committed:
lattice: remove unnecessary includes.

KaiSzuttor committed:
Formatting.

fweik commented on src/core/forces.cpp:
```diff
@@ -34,10 +34,10 @@
 #include "forcecap.hpp"
 #include "forces_inline.hpp"
 #include "grid_based_algorithms/electrokinetics.hpp"
+#include "grid_based_algorithms/lattice.hpp"
```
Why is this needed?

fweik commented on src/python/espressomd/system.pyx:
```diff
@@ -294,9 +294,10 @@ cdef class System(object):
-        def __get__(self):
-            return lattice_switch
+    IF LB or LB_GPU:
+        property lattice_switch:
```
Should this really be exposed? And if so why on the `System` class?

KaiSzuttor replied:
No, it shouldn't. You are right.

KaiSzuttor committed:
lattice: remove unnecessary include.

KaiSzuttor committed:
lattice: don't expose lattice_switch to the system class.

fweik commented:
Please check all includes, some are suspicious, like in `rotation.hpp`

KaiSzuttor commented:
I did

fweik commented:
```
core/global.cpp:#include "grid_based_algorithms/lattice.hpp"
core/forces.cpp:#include "grid_based_algorithms/electrokinetics.hpp"
core/forces.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/forces.cpp:#include "grid_based_algorithms/lb_particle_coupling.hpp"
core/rotation.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/communication.cpp:#include "grid_based_algorithms/lb.hpp"
core/communication.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/communication.cpp:#include "grid_based_algorithms/lb_interpolation.hpp"
core/communication.cpp:#include "grid_based_algorithms/lb_particle_coupling.hpp"
core/statistics.cpp:#include "grid_based_algorithms/lb.hpp"
core/statistics.cpp:#include "grid_based_algorithms/lbgpu.hpp"
python/espressomd/electrokinetics.pxd:    cdef extern from "grid_based_algorithms/electrokinetics.hpp":
script_interface/lbboundaries/LBBoundaries.hpp:#include "core/grid_based_algorithms/lbboundaries.hpp"
script_interface/lbboundaries/LBBoundary.hpp:#include "core/grid_based_algorithms/lbboundaries/LBBoundary.hpp"
python/espressomd/lb.pxd:    cdef extern from "grid_based_algorithms/lb_interface.hpp":
python/espressomd/lb.pxd:    cdef extern from "grid_based_algorithms/lb_particle_coupling.hpp":
python/espressomd/lb.pxd:    cdef extern from "grid_based_algorithms/lbgpu.hpp":
python/espressomd/lb.pxd:    cdef extern from "grid_based_algorithms/lb_interpolation.hpp":
core/integrate.cpp:#include "grid_based_algorithms/electrokinetics.hpp"
core/integrate.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/integrate.cpp:#include "grid_based_algorithms/lb_particle_coupling.hpp"
core/initialize.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/initialize.cpp:#include "grid_based_algorithms/lbboundaries.hpp"
core/thermostat.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/object-in-fluid/oif_global_forces.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/observables/LBVelocityProfile.cpp:#include "grid_based_algorithms/lattice.hpp"
core/observables/LBVelocityProfile.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/observables/LBVelocityProfile.cpp:#include "grid_based_algorithms/lb_interpolation.hpp"
core/virtual_sites/VirtualSitesInertialessTracers.cpp:#include "grid_based_algorithms/lattice.hpp"
core/virtual_sites/lb_inertialess_tracers_cuda.cu:#include "grid_based_algorithms/lbboundaries.hpp"
core/virtual_sites/lb_inertialess_tracers_cuda.cu:#include "grid_based_algorithms/lbgpu.cuh"
core/virtual_sites/lb_inertialess_tracers_cuda.cu:#include "grid_based_algorithms/lbgpu.hpp"
core/observables/CylindricalLBVelocityProfileAtParticlePositions.cpp:#include "grid_based_algorithms/lattice.hpp"
core/observables/CylindricalLBVelocityProfileAtParticlePositions.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/observables/CylindricalLBVelocityProfileAtParticlePositions.cpp:#include "grid_based_algorithms/lb_interpolation.hpp"
core/virtual_sites/lb_inertialess_tracers.cpp:#include "grid_based_algorithms/lb.hpp"
core/virtual_sites/lb_inertialess_tracers.cpp:#include "grid_based_algorithms/lbboundaries.hpp"
core/observables/CylindricalLBVelocityProfile.cpp:#include "grid_based_algorithms/lattice.hpp"
core/observables/CylindricalLBVelocityProfile.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/observables/CylindricalLBVelocityProfile.cpp:#include "grid_based_algorithms/lb_interpolation.hpp"
core/observables/CylindricalLBFluxDensityProfileAtParticlePositions.cpp:#include "grid_based_algorithms/lattice.hpp"
core/observables/CylindricalLBFluxDensityProfileAtParticlePositions.cpp:#include "grid_based_algorithms/lb_interface.hpp"
core/observables/CylindricalLBFluxDensityProfileAtParticlePositions.cpp:#include "grid_based_algorithms/lb_interpolation.hpp"
core/unit_tests/lb-d3q19.cpp:#include "grid_based_algorithms/lb-d3q19.hpp"
```

fweik commented:
e.g. why are
```
core/statistics.cpp:#include "grid_based_algorithms/lb.hpp"
core/statistics.cpp:#include "grid_based_algorithms/lbgpu.hpp"
```
needed?

RudolfWeeber reviewed (approved):
Huge improvement. Thanks alot for doing this work, Kai.

KaiSzuttor committed:
LB: remove non-interface include from statistics.

KaiSzuttor committed:
LB: remove lattice includes from observables.

KaiSzuttor committed:
Formatting.

KaiSzuttor commented:
I see no non-trivial way of removing the communication LB dependency but I cleaned up some of the other includes.

KaiSzuttor committed:
statistics: cleanup.

KaiSzuttor committed:
Formatting.
